pkg/deployment/autoscale.go

```diff
@@ -0,0 +1,107 @@
+package deployment
+
+import (
+	autoscalingv2beta2 "k8s.io/api/autoscaling/v2beta2"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1 "github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1"
+	"github.com/jaegertracing/jaeger-operator/pkg/util"
+)
+
+const (
+	// we need to have an upper bound, and 100 seems like a "good" max value
+	defaultMaxReplicas = int32(100)
+
+	// for both memory and cpu
+	defaultAvgUtilization = int32(90)
+)
+
+// Autoscalers returns a list of HPAs based on specs
+func autoscalers(
+	componentReplicas *int32,
+	label string,
+	componentName string,
+	componentLabels map[string]string,
+	autoScaleSpec v1.AutoScaleSpec,
+	componentSpec v1.JaegerCommonSpec,
+	jaeger *v1.Jaeger) []autoscalingv2beta2.HorizontalPodAutoscaler {
+
+	// fixed number of replicas is explicitly set, do not auto scale
+	if componentReplicas != nil {
+		return []autoscalingv2beta2.HorizontalPodAutoscaler{}
+	}
+
+	// explicitly disabled, do not auto scale
+	if autoScaleSpec.Autoscale != nil && *autoScaleSpec.Autoscale == false {
+		return []autoscalingv2beta2.HorizontalPodAutoscaler{}
+	}
+
+	maxReplicas := int32(-1) // unset, or invalid value
+
+	if nil != autoScaleSpec.MaxReplicas {
+		maxReplicas = *autoScaleSpec.MaxReplicas
+	}
+	if maxReplicas < 0 {
+		maxReplicas = defaultMaxReplicas
+	}
+
+	componentLabels["app.kubernetes.io/component"] = label
```

Comment:
Just wondering if this should be set here, when the `componentLabels` and `label` are both passed in and I don't think `label` is used anywhere else? So maybe better to add the component label before supplying just the `componentLabels` parameter?

Reply:
good observation, I'll do the change.

Reply:
mmm after thought more about this, I would prefer to have all of autoscaling functionality  encapsulated here, even if it's only one assignment.

Reply:
Ok not a big problem.